Replace the WYSIWYG editor with TipTap

ahukkanen opened the pull request:
#### :tophat: What? Why?
This PR implements the TipTap editor for Decidim.

This is feature complete and works both in the legacy design (admin panel) as well as in the redesign (you need to test it locally not to break staging, instructions below).

Can be tested at (use default seed credentials):
https://tiptap.decidimqa.mainiotech.fi/

#### :pushpin: Related Issues
- Related to
  * #7289
  * #7585
  * quilljs/quill#3359
- Fixes #10082
- Fixes #9398

Here are also some issues we have had in the past with Quill that should be confirmed to be fixed when testing this:

- [x] When saving and re-editing the content, sub-lists are loaded and displayed correctly
  #9866
- [x] Character counter does not count the HTML tags in the character count, see e.g. new proposal
  #9821
- [x] Copy-pasting content to the editor works
  #7180
  * [x] Works also for description field of assembly meetings, i.e. editors with hashtags
    #4873
- [x] Bold text is not lost when copy-pasting content from another editor
  #8982
- [x] Editor loads the embedded video content when that is the only content
  #8331
  #9629
- [x] Editor loads correctly when it contains paragraphs that only have a single line break inside them
  #7896
- [x] File upload works through the toolbar as well as copy-paste
  #7712
- [x] Does not create a trailing paragraph after a video element
  #3067
- [x] Videos display correctly when added to the pages
  #2130
  * [x] Note that you need to have accepted cookies for this to work
    #9170 (note: [not yet implemented under redesign](https://github.com/decidim/decidim/issues/9753#issuecomment-1369056254))
- [x] Pasting plain text content that has two line breaks between two text lines creates two separate paragraphs
  #2000
  * E.g. `First paragraph\n\nSecond paragraph`
- [x] Hashtags are recognized at the beginning of the content
  #9616
- [x] External link indicator is not added for content inside the editor
  #9095
- [x] Editor adds `nofollow`, `noopener` and `noreferrer` for participant entered content
  #9047
  #8779
  #5651
- [x] Character counter counts the emoji characters correctly, the same way as backend
  #8916
- [x] Shift+Enter makes a single line break instead of another paragraph
  #7877
- [x] External links are displayed as inline content when the WYSIWYG content is shown
  #7280
- [x] Alt tags are not removed for existing rich text content that has images with alt tags
  #6920

#### Testing
- Clone this branch, create the development app (you may need to regenerate it as `packages` have changed)
- In another window, open [try.decidim.org](https://try.decidim.org/)
- Go to admin panel in both versions
- Find a view with WYSIWYG editor
- Try it out
- Compare that we have all functionality available that we used to have with Quill

Currently there does not seem to be any redesigned view yet with the editor (e.g. new proposal / new meeting / new initiative / etc.) but you can test it locally under the redesign by adding it to some page, e.g. at [the static page view](https://github.com/decidim/decidim/blob/develop/decidim-core/app/views/decidim/pages/_tabbed.html.erb) as follows:

```erb
          <!-- Add this BELOW the <div class="editor-content mt-6">...</div> wrapper -->
          <div class="mt-6">
            <%= decidim_form_for(Decidim::StaticPage.new, url: "") do |form| %>
              <%= form.editor(:content) %>
            <% end %>
          </div>
```

And if you want to test the full toolbar with the video embed and image embed functionality, replace the line with `form.editor` with this (note that you need to be logged in for the image embedding to work because of the permissions set for the image upload endpoint):

```erb
<%= form.editor(:content, toolbar: "full") %>
```

Note that this PR already includes E2E tests (system tests) for testing the editor functionality under both, the legacy design as well as the redesign.

##### Extra features

The new editor also comes with some extra features that you can test:

- Automatically formatting pasted content
  * Try pasting a YouTube URL, e.g. `https://www.youtube.com/watch?v=f6JMgJAQ2tc`
  * Try pasting markdown content, e.g. `**this should appear as bold**`
- Markdown shortcuts
  * Try writing e.g. a heading to the editor, start the text with `##` + `SPACE` and continue writing your heading
  * Try writing e.g. a bold to the editor, write double stars `**` followed by some text and then again double stars `**`
- Character counter
  * You can try it at the participant side, e.g. when submitting a new proposal
- Hastags support, try e.g. with new proposal
- Mentions support (right now we do not have any WYSIWYG editor in Decidim that supports mentions but the support is built to the editor)
- Double clicking existing embedded content (images or videos) to edit them

### :camera: Screenshots

#### Participant editor with demonstration of the available controls

![Participant WYSIWYG](https://user-images.githubusercontent.com/864340/216444586-9b907fc6-2987-459b-860f-00b52ee774ae.png)

#### Admin editor with extra controls

![Admin editor view](https://user-images.githubusercontent.com/864340/216444750-6e87c3d6-3e90-453f-a53e-214fc784231a.png)

##### Link dialog

![WYSIWYG link dialog](https://user-images.githubusercontent.com/864340/216444829-385c8d89-8e09-4bf8-b384-2b96236ef7c9.png)

##### Image dialog

![WYSIWYG image dialog](https://user-images.githubusercontent.com/864340/216444942-743992b1-a985-42f9-8211-fb1e5988427b.png)

##### Video dialog

![WYSIWYG video dialog](https://user-images.githubusercontent.com/864340/216445216-3a6d27a4-9a87-49d1-955e-0097453555a7.png)

Supports YouTube and Vimeo by default or any arbitrary service as long as they provide embeddable video URLs.

##### Image resizing support

As in the old editor, we have image resizing support also in the new editor.

![WYSIWYG image resizing](https://user-images.githubusercontent.com/864340/216450341-3bc71106-1fd5-41ee-b67e-ae6a3a880618.png)

#### Redesigned editor

![Redesigned WYSIWYG](https://user-images.githubusercontent.com/864340/216447470-c99b51e3-e963-4cb6-976f-ab3bb1b63d5f.png)

##### Link dialog (redesigned)

![WYSIWYG link dialog (redesigned)](https://user-images.githubusercontent.com/864340/216447654-df0c3513-b723-42d1-9a3e-96e6f7bbe723.png)

##### Image dialog (redesigned)

![WYSIWYG image dialog (redesigned)](https://user-images.githubusercontent.com/864340/216448121-1af36afd-d38f-4e82-8fae-87e83ff73c44.png)

##### Video dialog (redesigned)

![WYSIWYG video dialog (redesigned)](https://user-images.githubusercontent.com/864340/216448011-79ee3305-6253-4f95-a08f-de70bb93075f.png)

alecslupu commented:
@ahukkanen This pr may fix also : #10060

ahukkanen commented:
@decidim/product This is now ready for acceptance testing.

There is still some minor tweaking to be done in few places but I'd say 99.5% done, so you can start testing it to report any issues you're able to find.

alecslupu reviewed (commented):
Just a partial code review. 

65 / 167 files viewed

ahukkanen commented on decidim-admin/app/packs/src/decidim/admin/tab_focus.js:
```diff
@@ -4,14 +4,16 @@
     // Detect if inside the tabs-panel have an input
     } else {
       $content = $container.find("input:first");
+      console.log($container);
```
Fixed by 085b7c2049d4290b49864767ca3d1035b90c5001

ahukkanen commented on decidim-admin/app/views/layouts/decidim/admin/_js_configuration.html.erb:
```diff
@@ -2,10 +2,8 @@
-      title: t("decidim.security.selfxss_warning.title"),
-      description: t("decidim.security.selfxss_warning.description")
-    }
+    "selfxssWarning": t("decidim.security.selfxss_warning"),
```
Yes, it is intended. Also similarly in the admin JS and redesigned file.

ahukkanen commented on docs/modules/customize/pages/javascript.adoc:
```diff
@@ -17,8 +17,8 @@ $(function(){
 
-If you want to replace a whole file existin in Decidim you can do it by creating a file in your application with the same name and the same path. I.e: to replace `decidim-core/app/packs/src/decidim/editor.js` you should create in your Rails app a file in
-  `app/packs/src/decidim/editor.js` and it will have more priority over the Decidim file.
+If you want to replace a whole file existin in Decidim you can do it by creating a file in your application with the same name and the same path. I.e: to replace `decidim-core/app/packs/src/decidim/editor/index.js` you should create in your Rails app a file in
```
Fixed by cebb3ede519f8fde17f1ba5ddb7eef29f8613df1

ahukkanen commented on decidim-meetings/spec/presenters/decidim/meetings/meeting_presenter_spec.rb:
```diff
@@ -99,5 +99,47 @@ module Decidim::Meetings
       end
     end
+
+    describe "#editor_description" do
```
Refactored in 466aecd0cc0a1513ffd0e18a641b5b6582ce853a

ahukkanen commented on decidim-forms/app/packs/src/decidim/forms/admin/forms.js:
```diff
@@ -387,8 +386,8 @@ export default function createEditableForm() {
 
-      $field.find(".editor-container").each((idx, el) => {
-        createQuillEditor(el);
+      $field[0].querySelectorAll(".editor-container").forEach((el) => {
```
The editor initialization has to happen in the correct context. In forms you can add fields dynamically so it has to happen at the form JS code. I don't think this is too much duplication.

ahukkanen commented on decidim-system/app/packs/src/decidim/system/application.js:
```diff
@@ -20,8 +18,8 @@ import "src/decidim/confirm"
 
-  $(".editor-container").each((_idx, container) => {
-    createQuillEditor(container);
+  document.querySelectorAll(".editor-container").forEach((container) => {
```
This does not belong to `decidim_editor.js`. That entrypoint is in charge of making the editor available for the front-end. Each context initializes the editor with the information they need in each context.

ahukkanen commented on decidim-admin/spec/system/admin_manages_organization_spec.rb:
```diff
@@ -40,7 +40,15 @@
         visit decidim_admin.edit_organization_path
 
         # Makes sure in the error screenshots the editor is visible
```
These tests have been already there in the past. I am just making the necessary modifications to these with the new editor.

These are some issues we have noticed with the editor in the past and I believe they are tested on purpose in the actual context where they live at. So that we are not re-introducing some of those bugs.

The `editor_spec` is outside of this context, it's a dummy view for testing the editor functionality. But I think it makes sense to also test it in the actual context, as we used to (particularly regarding those bugs, so that we are not re-introducing them).

ahukkanen commented on decidim-comments/spec/events/decidim/comments/user_group_mentioned_event_spec.rb:
```diff
@@ -8,7 +8,7 @@
   let(:ca_comment_content) { "<div><p>Un commentaire pour #{author_link}</p></div>" }
   let(:en_comment_content) { "<div><p>Comment mentioning some user group, #{author_link}</p></div>" }
-  let(:author_link) { "<a class=\"user-mention\" href=\"http://#{organization.host}:#{Capybara.server_port}/profiles/#{group.nickname}\">@#{group.nickname}</a>" }
+  let(:author_link) { "<a href=\"http://#{organization.host}:#{Capybara.server_port}/profiles/#{group.nickname}\" class=\"user-mention\" target=\"_blank\" rel=\"nofollow noopener noreferrer ugc\">@#{group.nickname}</a>" }
```
Well, not necessarily in the scope but since we are touching the `LinkRenderer` in this PR (which is necessary), this is also affected by that change.

ahukkanen commented on decidim-comments/spec/events/decidim/comments/user_mentioned_event_spec.rb:
```diff
@@ -10,7 +10,7 @@
   let(:ca_comment_content) { "<div><p>Un commentaire pour #{author_link}</p></div>" }
   let(:en_comment_content) { "<div><p>Comment mentioning some user, #{author_link}</p></div>" }
-  let(:author_link) { "<a class=\"user-mention\" href=\"http://#{organization.host}:#{Capybara.server_port}/profiles/#{author.nickname}\">@#{author.nickname}</a>" }
+  let(:author_link) { "<a href=\"http://#{organization.host}:#{Capybara.server_port}/profiles/#{author.nickname}\" class=\"user-mention\" target=\"_blank\" rel=\"nofollow noopener noreferrer ugc\">@#{author.nickname}</a>" }
```
Well, not necessarily in the scope but since we are touching the `LinkRenderer` in this PR (which is necessary), this is also affected by that change.

carolromero reviewed (changes requested):
Fantastic work as usual @ahukkanen! :tada:
Here is the feedback of the things we found during the review:

### Paste styles from another document
In general it works great! But when pasting lists with sub-items it loses the original style.

![image](https://user-images.githubusercontent.com/23509895/217303815-59442ad4-1438-4c26-bfee-7831952c18e1.png)
![image](https://user-images.githubusercontent.com/23509895/217304285-1497757f-2ee8-4b36-ba65-98635aeb78f0.png)

### Embedding an image
Images are embedded in all fields, whether they are copied/pasted or dragged directly into the editor. However, the image dialog only works on drop, not on click. 

[dragndrop_image.webm](https://user-images.githubusercontent.com/23509895/217307944-dfbfecd4-194e-42e8-9cbd-e57ea333d6fe.webm)

### Resizing an image
Doesn’t work in Firefox (109.0.1 (64 bits)) - Does in Chrome.

### Embedding a video
It works correctly in the description field of a process or assembly. In **the rest of the fields it is embedded but it is not shown**. In the component editor it is also embedded but not shown.

**Admin view**
![image](https://user-images.githubusercontent.com/23509895/217311158-83aa7527-aa53-47e7-8ec0-0eb955705329.png)

**Frontend view**
![image](https://user-images.githubusercontent.com/23509895/217310959-4d4d5d25-bfac-42c3-b8d0-84f6f89bee35.png)

I have checked that this is currently happening with Quill, I don't know if it is related to this issue? https://github.com/decidim/decidim/issues/10130

### Links
When inserting a link from the frontend (for instance in a debate by a participant) it doesn't save the setting that it opens in the same tab.

[sametablink.webm](https://user-images.githubusercontent.com/23509895/217313744-c65ed6c3-4019-458e-b43c-75964a04c4e6.webm)

From the admin it works fine, but if you copy a text with links that open in the same tab, when pasted it changes the settings to a new tab.

### Hashtags

**With TipTap**
![image](https://user-images.githubusercontent.com/23509895/217315209-3d13df67-5bcd-48ec-89e9-f49a26f07e1f.png)

**With Quill**
![image](https://user-images.githubusercontent.com/23509895/217315330-9583205c-4fe8-42d3-a466-5cc112a05e6a.png)

ahukkanen commented:
Thank you very much for your review @carolromero!

The discovered issues have been fixed and I'd like to ask you to have a second run to confirm whether all of these are OK. Please also see some explanations below for the behavior.

The staging instance has been updated with the latest changes.

---

> ### Paste styles from another document
> In general it works great! But when pasting lists with sub-items it loses the original style.

This has now been fixed regarding the sub-list styles, i.e. preserving the `a.`, `b.`, `c.` list styles on the sub-lists. Please note, however, that HTML does not support all the same styles provided by the desktop editors. E.g. all styles `(a).`, `a).` and `a.` will convert to the same style on the web i.e. `a.`.

This was actually a bug in Tiptap which I reported also at their repository: ueberdosis/tiptap#3726

Regarding the bold content in your screenshot, I am not able to replicate this issue. For me, the content in the sub-lists remains with the normal style. In case you are still experiencing that issue, please send the document that you are trying to paste from and we'll have a closer look.

> ### Embedding an image
> Images are embedded in all fields, whether they are copied/pasted or dragged directly into the editor. However, the image dialog only works on drop, not on click.

This same issue was discovered in our internal testing by @JoonasAapro. He also went ahead and fixed it already.

> ### Resizing an image
> Doesn’t work in Firefox (109.0.1 (64 bits)) - Does in Chrome.

This bug is now fixed.

> ### Embedding a video
> It works correctly in the description field of a process or assembly. In **the rest of the fields it is embedded but it is not shown**. In the component editor it is also embedded but not shown.
> 
> [... screenshots removed ...]
>
> I have checked that this is currently happening with Quill, I don't know if it is related to this issue? #10130

Yes, this is related to #10130, it's a separate issue. It has something to do how we are pre-processing the content that gets shown in different places.

But it does have something to do with the editor as well, as we discussed earlier. In these cases, we should not allow embedding images or videos through the editor.

Either way, it should be considered as a separate issue because this is also the current behavior we have with Quill (I believe).

> ### Links
> When inserting a link from the frontend (for instance in a debate by a participant) it doesn't save the setting that it opens in the same tab.
> 
> From the admin it works fine, but if you copy a text with links that open in the same tab, when pasted it changes the settings to a new tab.

This is kind of by design because we want to force the participant entered links to open externally so that they cannot trick users clicking malicious links that would leak the user information.

What we have done is that we removed the target selector from the link dialog for participants. Only admins can control the link target now.

> ### ~Hashtags~ (actually this is not happening with the editor)

Note that many things in the participant facing interface may seem a bit broken due to the ongoing redesign process. The redesign work is breaking some of the old views after some of the components visible in those views have been already redesigned.

We just cannot enable the redesigned layout in the staging app because it is not finished yet, so we would break many more things if we enable redesign right now.

carolromero commented:
Thank you @ahukkanen for the quick fix and the explanations! A couple of issues that persist:

> ## Paste styles from another document
> In general it works great! But when pasting lists with sub-items it loses the original style.
> 
> This has now been fixed regarding the sub-list styles, i.e. preserving the a., b., c. list styles on the sub-lists. Please note, however, that HTML does not support all the same styles provided by the desktop editors. E.g. all styles (a)., a). and a. will convert to the same style on the web i.e. a..
> 
> This was actually a bug in Tiptap which I reported also at their repository: https://github.com/ueberdosis/tiptap/issues/3726
> 
> Regarding the bold content in your screenshot, I am not able to replicate this issue. For me, the content in the sub-lists remains with the normal style. In case you are still experiencing that issue, please send the document that you are trying to paste from and we'll have a closer look.
> 

Oh I see. Indeed, from LibreOffice it is working fine. The document I'm using is from Google Docs: https://docs.google.com/document/d/1sbi0GZlEWB08DVYjxA2jf6Xt0cJpseEWuotKbrugUDM/edit?usp=sharing

> ## Links
> When inserting a link from the frontend (for instance in a debate by a participant) it doesn't save the setting that it opens in the same tab.
> 
> From the admin it works fine, but if you copy a text with links that open in the same tab, when pasted it changes the settings to a new tab.
> 
> This is kind of by design because we want to force the participant entered links to open externally so that they cannot trick users clicking malicious links that would leak the user information.
> 
> What we have done is that we removed the target selector from the link dialog for participants. Only admins can control the link target now.

I noticed that now the link can no longer be edited (I'd swear that when I checked it the other day I was able to do it several times). Can we get it back?

[editlink.webm](https://user-images.githubusercontent.com/23509895/217816597-1c53e7b7-6d86-41fc-afc9-21cd73640968.webm)

alecslupu commented on decidim-core/lib/decidim/content_processor.rb:
```diff
@@ -150,5 +150,30 @@ def self.parser_klass(type)
+      #
+      # @return [Loofah::HTML::DocumentFragment]
+      def html_fragment(text = nil)
+        if text
```
Maybe we could simplify to some like (and avoid multiple parsings when we have text):  
```
@fragment ||= Loofah.fragment(text || content)
```

ahukkanen replied:
No. That would always store the first content passed to the method meaning this would change the behavior and break things.

Only the content fragment is stored but any arbitrary text passed to this method should not be stored / would not make sense to store.

alecslupu reviewed (commented):
Looks good to me.

alecslupu commented:
@ahukkanen I have reviewed most of the work, and i cannot say that i have any other comments of improvements on the PR. 
I have not tested yet from the functionality point of view (generics), and I am expecting @carolromero to finalize her product review.

carolromero reviewed (approved):
@ahukkanen thanks! I love the solution you found for editing links, beautifully solved.
Yay, we can already say that we have a new WYSIWYG editor and it's definitely better than the old one!

alecslupu reviewed (commented):
LGTM.

ahukkanen commented:
@alecslupu This is now finally ready for technical review. I have done some further testing, some refactoring and added more tests for the added features / bugs that were discovered during the product review.

---

Few notes regarding the latest changes (about copy-pasting from external editors):

While testing copying content from the various editors, I bumped into other issues regarding pasting lists. Particularly I was testing Office 365 and desktop Word. All of the document editors seem to be producing quite funky HTML and require specific processing for each editor.

Now the following editors should work fine at least when pasting lists:
- LibreOffice Writer
- Google Docs
- Office 365 Word (web version)
- Desktop version of Microsoft Word

Especially the Microsoft editors required very specific processing for their markup. I have added some content processing code to the Decidim editor implementation that I feel should not live in the Decidim repository. I have reported the following issues to the Tiptap repository regarding these issues concerning the lists in copy-pasted content:
- ueberdosis/tiptap#3726
- ueberdosis/tiptap#3735
- ueberdosis/tiptap#3751
- ueberdosis/tiptap#3756

I would hope that we would eventually get some of this processing code moved to Tiptap because IMO this kind of stuff belongs to the editor itself.

Additionally, I have also discovered an issue regarding the character counter which I have also reported to Tiptap repository: ueberdosis/tiptap#3721. This is also already fixed in the Decidim implementation through a custom extension.

ahukkanen commented:
And also to note about styling the ordered lists, I discovered another issue with Tailwind that I reported at tailwindlabs/tailwindcss-typography#296.

Tailwind overrides the browser's default styling for the "typed" ordered lists which causes an issue with their styling. This is because elements cannot be styled through CSS using the `type` attribute in case we have to target the values of those attributes case sensitively i.e. (`a` != `A`). 

This is an issue with CSS itself (see whatwg/html#4158). Fortunately the CSS working group accepted a case sensitive attribute selector modifier to the spec (see w3c/csswg-drafts#2101) but the unfortunate part is that this modifier is not yet widely implemented in all major browsers:
https://caniuse.com/mdn-css_selectors_attribute_case_sensitive_modifier

While we are waiting the feature to be shipped in all major browsers, I have added another workaround for this issue in the Decidim editor implementation in this commit:
https://github.com/decidim/decidim/pull/10196/commits/c4da43bd51db413d9dc823f83cbf45266e3fcf1a

Once the case sensitive `s` modifier becomes more widely implemented in browsers, I would expect we can remove the changes done in that commit, as I would expect Tailwind to fix this issue. Until then, we need to keep these changes as a workaround for that issue.

alecslupu commented:
@ahukkanen , there are some conflicts on this branch. Could you fix'em and request review again ?

alecslupu commented on decidim-core/app/packs/src/decidim/editor/common/input_dialog.js:
```diff
@@ -0,0 +1,214 @@
+    const i18n = getDictionary("editor.inputDialog");
+
+    let buttonsHTML = "";
+    if (this.legacyDesign) {
```
Why do we have a check here, to treat the legacy Design, yet  exactly 8 lines lower you do the same thing ? Could we have one single check combining the 2 ?

ahukkanen replied:
Good point, there was something else here before which is why it was like this.

Fixed at 9187ee5c3005d7d36574878f11642d73e7b1af91.

alecslupu reviewed (changes requested):
Please solve the conflicts, and see the comment i have left behind

alecslupu commented:
I am reviewing from the client perspective, and i have found a small issue while attaching pictures. It seems that when I try to resize the embedded picture, there is a form submit, instead of actual resize. 

Additionally, i have tried to embed a Youtube video having a link like https://www.youtube.com/watch?v=J8Xwr1kTEdk , and the element is not being rendered in the page, on the proposals body, instead being rendered a blank div like 
```
<div class="editor-content-videoEmbed"><div></div></div>
```

alecslupu reviewed (changes requested):
I am reviewing from the client perspective, and i have found a small issue while attaching pictures. It seems that when I try to resize the embedded picture, there is a form submit, instead of actual resize. 

Additionally, i have tried to embed a Youtube video having a link like https://www.youtube.com/watch?v=J8Xwr1kTEdk , and the element is not being rendered in the page, on the proposals body, instead being rendered a blank div like 
```
<div class="editor-content-videoEmbed"><div></div></div>
```

ahukkanen commented:
The lastest bug regarding form submit on image resize is fixed by a176b5f321c8ad79d1107320f7cd21aec4f90948. I also added specs to cover this case.

Regarding the other bug that the video embed is not displayed, this is sort of "by design" because we are stripping the multimedia (embedded iframes and images) from the content potentially entered by participants. It is another topic whether we should allow iframes and images for admin entered proposals, i.e. "official" proposals.

This issue is further explained at #10130 and we should handle that separately. I'll add some details to that issue later today.

As discussed in the maintainers meeting, let's keep this PR on hold until that issue is resolved. Otherwise the fix may be hard to backport.

ahukkanen commented:
@decidim/maintainers This is now finally ready for technical review after the current editor toolbar issues have been resolved at `develop` and the changes have been reflected to this branch.

Meanwhile we have been waiting to merge this, we have also received a stable release last month for Tiptap version 2 :partying_face::
https://tiptap.dev/blog/tiptap-release-day

I have also updated the Tiptap to the stable version.

andreslucena commented:
>Meanwhile we have been waiting to merge this, we have also received a stable release last month for Tiptap version 2 🥳 :
[tiptap.dev/blog/tiptap-release-day](https://tiptap.dev/blog/tiptap-release-day)
>
> I have also updated the Tiptap to the stable version.

Great news!! 👏🏽 👏🏽 

> EDIT: Staging/QA environment has also been updated with the latest code.

@carolromero could you give it an eye? Thanks!

alecslupu reviewed (commented):
Partial Reviewed, I have a quill related issue while upgrading.

alecslupu commented on decidim-core/app/packs/src/decidim/input_character_counter.js:
```diff
@@ -264,7 +263,7 @@ export default class InputCharacterCounter {
         message = MESSAGES.charactersLeft.one;
       }
-      this.$input[0].dispatchEvent(
+      this.$userInput[0].dispatchEvent(
```
I have noticed that we have `this.$input` and `this.$userInput` (maybe the same we have defined in the line 92). what is the difference between those 2 calls?

ahukkanen replied:
The first one (`this.$input`) refers to the plain input field, which in normal situations would be e.g. input or textarea.

The other one (`this.$userInput`) refers to the contenteditable area, in this case the ProseMirror controlled div. The value of the contenteditable area is not sent to the server but it is added to the hidden input field (i.e. `this.$input`) when the contenteditable area is updated.

In case of a rich text editor, `this.$input` variable refers to the hidden input field containing the value to be sent to the server and `this.$userInput` refers to the contenteditable area. We want to dispatch the event directly to the contenteditable area so that the editor itself can be completely unaware of the hidden input related to it.

alecslupu commented on decidim-core/app/packs/src/decidim/redesigned_input_character_counter.js:
```diff
@@ -269,7 +268,7 @@ export default class InputCharacterCounter {
         message = MESSAGES.charactersLeft.one;
       }
-      this.$input[0].dispatchEvent(
+      this.$userInput[0].dispatchEvent(
```
please see comment in : decidim-core/app/packs/src/decidim/input_character_counter.js

ahukkanen replied:
Same as above.

alecslupu commented on decidim-core/spec/system/editor_spec.rb:
```diff
@@ -0,0 +1,1768 @@
+      end
+    end
+
+    def drag(selector, mode: "mouse", direction: nil, amount: 0)
```
Nice implementation. Maybe it would better to extract this, or a derivate function as stand alone ? It is the kind of method that would allow us to write specs for hompage content block ( reorder ) functionality.

carolromero commented:
> @carolromero could you give it an eye? Thanks!

I took a look and everything seems to be fine :+1:

ahukkanen commented on RELEASE_NOTES.md:
```diff
@@ -46,7 +46,21 @@ You will need to add `tailwind.config.js` to your app `.gitignore`. If you gener
 You can read more about this change on PR [\#9480](https://github.com/decidim/decidim/pull/9480).
 
-### 3.2. Added Procfile support
+### 3.2. Content migration for rich text editor
```
This is fixed by b3dc89b559c86d184af5100cff44cb34f426f932.

The `decidim:upgrade` task needs to be run on upgrade as part of the generic release notes at the top. I added there copying the webpack configurations which fixes this issue.

alecslupu commented on decidim-core/lib/tasks/upgrade/decidim_migrate_wysiwyg_content.rake:
```diff
@@ -0,0 +1,91 @@
+      # Core
+      Decidim::Upgrade::WysiwygMigrator.register_model(
+        "Decidim::Organization",
+        [:description, :omnipresent_banner_short_description, :highlighted_content_banner_short_description, :welcome_notification_body, :admin_terms_of_use_body]
```
```suggestion
        [:description, :omnipresent_banner_short_description, :highlighted_content_banner_short_description, :welcome_notification_body, :admin_terms_of_service_body]
```

ahukkanen replied:
This is fixed with the latest changes. I also added specs for the migration rake tasks.

alecslupu reviewed (changes requested):
As per documentation, the task `bin/rails decidim:upgrade:migrate_wysiwyg_content` should migrate all the content . However, when i run the task, i get the following error: 

```
ActiveRecord::StatementInvalid: PG::UndefinedColumn: ERROR:  column "admin_terms_of_use_body" does not exist
LINE 1: ...cidim_organizations"."welcome_notification_body", "admin_ter...
                                                             ^
/code/decidim/decidim-core/lib/decidim/upgrade/wysiwyg_migrator.rb:48:in `block in convert_model_data'
/code/decidim/decidim-core/lib/decidim/upgrade/wysiwyg_migrator.rb:47:in `convert_model_data'
/code/decidim/decidim-core/lib/decidim/upgrade/wysiwyg_migrator.rb:32:in `block in update_records_batches'
/code/decidim/decidim-core/lib/decidim/upgrade/wysiwyg_migrator.rb:31:in `each_with_index'
/code/decidim/decidim-core/lib/decidim/upgrade/wysiwyg_migrator.rb:31:in `update_records_batches'
/code/decidim/decidim-core/lib/decidim/upgrade/wysiwyg_migrator.rb:80:in `block in update_models'
/code/decidim/decidim-core/lib/decidim/upgrade/wysiwyg_migrator.rb:79:in `each'
/code/decidim/decidim-core/lib/decidim/upgrade/wysiwyg_migrator.rb:79:in `update_models'
/code/decidim/decidim-core/lib/tasks/upgrade/decidim_migrate_wysiwyg_content.rake:53:in `block (3 levels) in <main>'
/code/bin/rails:5:in `<top (required)>'
/code/bin/spring:10:in `block in <top (required)>'
/code/bin/spring:7:in `<top (required)>'
```

It seems that in the task  ( decidim:upgrade:register_wysiwyg_migration) the old field name ( admin_terms_of_use_body ) is being used. As per change done in #10614, the field should be renamed to `admin_terms_of_service_body`.

After this change is being made, all should be ok from my end.
